File: bridge_iq/logger.py
"""
Logging configuration for the BridgeIQ client.

This module provides a customized logger for the client library that
supports different log levels and formats.
"""
import logging
import os
import sys
from logging.handlers import RotatingFileHandler
from pathlib import Path
from typing import Optional, Union

import platformdirs


def setup_logger(
    name: str = "bridge_iq",
    level: Union[int, str] = logging.INFO,
    log_to_file: bool = False,
    log_file: Optional[Union[str, Path]] = None,
    max_file_size: int = 10 * 1024 * 1024,  # 10 MB
    backup_count: int = 3,
    format_string: Optional[str] = None,
) -> logging.Logger:
    """Set up a logger with console and optional file handlers.
    
    Args:
        name: Logger name
        level: Logging level (DEBUG, INFO, WARNING, ERROR, CRITICAL)
        log_to_file: Whether to log to a file
        log_file: Path to log file. If None, uses default location
        max_file_size: Maximum size of log file before rotation (bytes)
        backup_count: Number of backup log files to keep
        format_string: Custom format string for log messages
        
    Returns:
        Configured logger instance
    """
    # Create logger
    logger = logging.getLogger(name)
    
    # Only set up handlers if none exist
    if not logger.handlers:
        # Set level
        if isinstance(level, str):
            level = getattr(logging, level.upper())
        logger.setLevel(level)
        
        # Create formatter
        if format_string is None:
            format_string = (
                "%(asctime)s - %(name)s - %(levelname)s - "
                "%(filename)s:%(lineno)d - %(message)s"
            )
        formatter = logging.Formatter(format_string)
        
        # Console handler
        console_handler = logging.StreamHandler(sys.stdout)
        console_handler.setFormatter(formatter)
        logger.addHandler(console_handler)
        
        # File handler (optional)
        if log_to_file:
            if log_file is None:
                # Use default location
                app_name = "bridge_iq"
                app_author = "thakaamed"
                log_dir = platformdirs.user_log_dir(app_name, app_author)
                os.makedirs(log_dir, exist_ok=True)
                log_file = Path(log_dir) / "bridge_iq.log"
            else:
                log_file = Path(log_file)
                os.makedirs(log_file.parent, exist_ok=True)
            
            file_handler = RotatingFileHandler(
                log_file,
                maxBytes=max_file_size,
                backupCount=backup_count,
            )
            file_handler.setFormatter(formatter)
            logger.addHandler(file_handler)
    
    return logger


# Default logger
logger = setup_logger()


def get_logger(
    name: Optional[str] = None,
    level: Optional[Union[int, str]] = None,
    log_to_file: bool = False,
) -> logging.Logger:
    """Get a configured logger.
    
    Args:
        name: Logger name. If None, returns the default logger
        level: Logging level. If None, uses the default level
        log_to_file: Whether to log to a file
        
    Returns:
        Configured logger instance
    """
    if name is None:
        # Return the default logger
        if level is not None:
            if isinstance(level, str):
                level = getattr(logging, level.upper())
            logger.setLevel(level)
        
        return logger
    
    # Create a new logger
    return setup_logger(
        name=name,
        level=level or logging.INFO,
        log_to_file=log_to_file,
    ) 

File: src/logger.ts
/**
 * Logging configuration for the BridgeIQ client.
 *
 * This module provides a customized logger for the client library that
 * supports different log levels and formats.
 */
import fs from 'fs';
import path from 'path';
import os from 'os';

/**
 * Log levels enum.
 */
export enum LogLevel {
  ERROR = 0,
  WARN = 1,
  INFO = 2,
  DEBUG = 3,
}

/**
 * Logger class for the Bridge IQ client.
 */
export class Logger {
  private level: LogLevel;
  private name: string;
  private logToFile: boolean;
  private logFile?: string;
  private logStream?: fs.WriteStream;

  /**
   * Create a new Logger.
   *
   * @param name - Logger name
   * @param level - Logging level (ERROR, WARN, INFO, DEBUG)
   * @param logToFile - Whether to log to a file
   * @param logFile - Optional path to log file
   */
  constructor(
    name = 'bridge_iq',
    level: LogLevel = LogLevel.INFO,
    logToFile = false,
    logFile?: string
  ) {
    this.name = name;
    this.level = level;
    this.logToFile = logToFile;

    if (logToFile) {
      // Set up logging to file
      if (!logFile) {
        // Use default location
        const logDir = path.join(os.homedir(), '.bridge_iq', 'logs');

        // Create log directory if it doesn't exist
        if (!fs.existsSync(logDir)) {
          fs.mkdirSync(logDir, { recursive: true });
        }

        this.logFile = path.join(logDir, 'bridge_iq.log');
      } else {
        this.logFile = logFile;

        // Create parent directory if it doesn't exist
        const logDir = path.dirname(this.logFile);
        if (!fs.existsSync(logDir)) {
          fs.mkdirSync(logDir, { recursive: true });
        }
      }

      // Open log file stream
      this.logStream = fs.createWriteStream(this.logFile, { flags: 'a' });
    }
  }

  /**
   * Format log message with timestamp and metadata.
   *
   * @param level - Log level label
   * @param message - Log message
   * @returns Formatted log message
   */
  private formatMessage(level: string, message: string): string {
    const timestamp = new Date().toISOString();
    return `${timestamp} - ${this.name} - ${level} - ${message}`;
  }

  /**
   * Write log message to console and optionally to file.
   *
   * @param level - Log level label
   * @param levelValue - Numeric log level
   * @param message - Log message
   */
  private log(level: string, levelValue: LogLevel, message: string): void {
    if (levelValue <= this.level) {
      const formattedMessage = this.formatMessage(level, message);

      // Log to console
      switch (levelValue) {
        case LogLevel.ERROR:
          console.error(formattedMessage);
          break;
        case LogLevel.WARN:
          console.warn(formattedMessage);
          break;
        case LogLevel.INFO:
          console.info(formattedMessage);
          break;
        case LogLevel.DEBUG:
          console.debug(formattedMessage);
          break;
      }

      // Log to file if enabled
      if (this.logToFile && this.logStream) {
        this.logStream.write(formattedMessage + '\n');
      }
    }
  }

  /**
   * Log a debug message.
   *
   * @param message - Log message
   */
  debug(message: string): void {
    this.log('DEBUG', LogLevel.DEBUG, message);
  }

  /**
   * Log an info message.
   *
   * @param message - Log message
   */
  info(message: string): void {
    this.log('INFO', LogLevel.INFO, message);
  }

  /**
   * Log a warning message.
   *
   * @param message - Log message
   */
  warn(message: string): void {
    this.log('WARN', LogLevel.WARN, message);
  }

  /**
   * Alias for warn() method.
   *
   * @param message - Log message
   */
  warning(message: string): void {
    this.warn(message);
  }

  /**
   * Log an error message.
   *
   * @param message - Log message
   */
  error(message: string): void {
    this.log('ERROR', LogLevel.ERROR, message);
  }

  /**
   * Close the logger and any open file streams.
   */
  close(): void {
    if (this.logStream) {
      this.logStream.end();
      this.logStream = undefined;
    }
  }
}

// Default logger instance
export const defaultLogger = new Logger();

/**
 * Get a configured logger instance.
 *
 * @param name - Logger name
 * @param level - Log level
 * @param logToFile - Whether to log to file
 * @param logFile - Optional log file path
 * @returns Logger instance
 */
export function getLogger(
  name?: string,
  level: LogLevel = LogLevel.INFO,
  logToFile = false,
  logFile?: string
): Logger {
  if (!name) {
    // Return default logger if no name provided
    defaultLogger.debug('Returning default logger');
    return defaultLogger;
  }

  // Create a new logger with the specified configuration
  return new Logger(name, level, logToFile, logFile);
}
